Automattic/woocommerce-payments
# Fix changelog for pre-release workflow when running from the release branch

Fixes #6461 

#### Changes proposed in this Pull Request

- Handle better the use case of `x.y.z-test-3` test package creation following #6309 
  - Theoretically, we now run the workflow from the release branch `release/x.y.z`
  - But this branch should already have [generated](https://github.com/Automattic/woocommerce-payments/blob/40b94fabe4efb919c72dfe5e2809c3497170dd4e/.github/workflows/release-pr.yml#L79) the changelog 

In such case, the solution is to replace the final version (`x.y.z`) by the test version (`x.y.z-test-3`) in the changelog (`changelog.txt` and `readme.txt`). To detect such case, we check whether the changelog folder is empty (i.e. only 1 `.gitkeep` file) and the release version provided (`x.y.z`) matches the release version defined in the `package.json` file.

As a bonus, I also decided to commit/push the changes made during the pre-release workflow (changelog + bump version in the plugin file) to have a trace later. Consequently, I moved the tag creation a bit later in the workflow to ensure we tag that commit.

#### Testing instructions

On a branch based on `fix/changelog-pre-release-workflow`:

* Run `vendor/bin/changelogger write --use-version="6.0.0" --no-interaction --yes` to generate the changelog in `changelog.txt`
* Copy/paste the 6.0.0 changelog in `readme.txt`
* Change the version in `package.json` to `6.0.0`

Now, you're ready to test the use case

- Run the [Create a GH pre-release](https://github.com/Automattic/woocommerce-payments/actions/workflows/create-pre-release.yml) on your test branch, using `6.0.0-test-4` as the release version
- Confirm that the changelog is updated correctly
  - You can check `changelog.txt` and `readme.txt` in the `testing/6.0.0-test-4` branch
  - You can check the GitHub release created for `6.0.0`: the description should show the changelog, and the zip associated to it should reflect the changes in the `testing/6.0.0-test-4` branch

-------------------


**Post merge**
